File: src/dispatcher/task.ts
import { TaskInfo, TaskResult } from './protocol/task';
import { NOOP } from './utils/noop';

export enum TaskStatus {
  Initial,
  PushedIntoQueue,
  PulledByDispatcher,
  RejectedByDispatcher,
  SetToProvider,
  SentFailed,
  SentToProvider,
  FailedByProvider,
  Aborted,
  Completed,
  Timeout
}

export type TaskStatusPayload = {
  [TaskStatus.SetToProvider]: [{ providerId: string, minScore: number, waitingTime: number}];
  [TaskStatus.SentFailed]: [{ attempt: number }];
  [TaskStatus.FailedByProvider]: [{ reason?: string }];
} & {
  [taskStatus: number | string | symbol]: [];
};

export class Task {
  private failedAttempts = 0;

  private priority = 0;

  private providerId: string | undefined = undefined;

  readonly status: TaskStatus = TaskStatus.Initial;

  private log: [Date, TaskStatus, TaskStatusPayload[TaskStatus]][] = [];

  constructor(readonly task_info: TaskInfo) {}

  get id(): string {
    return this.task_info.id;
  }

  getMaxPrice(): number {
    return this.task_info.max_price;
  }

  /** Returns time to money ratio */
  getMoneyTimeRatio(): number {
    return this.task_info.time_to_money_ratio;
  }

  addFailedAttempt(): void {
    this.failedAttempts++;
  }

  getFailedAttempts(): number {
    return this.failedAttempts;
  }

  // Priority queue related methods

  setPriority(priority: number): void {
    this.priority = priority;
  }

  getPriority(): number {
    return this.priority;
  }

  // Profiling related methods

  setStatus<T extends TaskStatus>(
    event: T,
    ...payload: TaskStatusPayload[T]
  ): void {
    // TODO:
    // We can add any profiling logic here, to analyze life cycle of a task

    this.log.push([new Date(), event, payload]);
  }

  getLog(): typeof this.log {
    return this.log;
  }

  getLogString(): string {
    return this.log.map(
      entry =>
        `[${entry[0].toISOString()}] ${TaskStatus[entry[1]]}: ${entry[2]
          .map(v => JSON.stringify(v))
          .join(', ')}`
    ).join('\n');
  }

  setProviderId(providerId: string): void {
    this.providerId = providerId;
  }

  getProviderId(): string | undefined {
    return this.providerId;
  }

  complete(result: TaskResult): void {
    this.onCompleted(result);
  }

  fail(): void {
    this.onFailed();
  }

  onCompleted: (result: TaskResult) => void = NOOP;
  onFailed: () => void = NOOP;
}

File: backend/src/dispatcher/task.py
from typing import Dict, Union, Tuple
from datetime import datetime
from dispatcher.utils.noop import NOOP
import json

class TaskStatus:
    Initial = 0
    PushedIntoQueue = 1
    PulledByDispatcher = 2
    RejectedByDispatcher = 3
    SetToProvider = 4
    SentFailed = 5
    SentToProvider = 6
    FailedByProvider = 7
    Aborted = 8
    Completed = 9
    Timeout = 10

TaskStatusPayload = Dict[
    Union[
      TaskStatus.SetToProvider,
      TaskStatus.SentFailed,
      TaskStatus.FailedByProvider
    ],
    Tuple[Dict[str, Union[str, int]], ...]
] 

class Task:
	def init(self, task_info):
		self.failedAttempts = 0
		self.priority = 0
		self.providerId = None
		self.status = TaskStatus.Initial
		self.log = []

		self.task_info = task_info

	def addFailedAttempt(self):
		self.failedAttempts += 1

	def getFailedAttempts(self):
		return self.failedAttempts

	def setPriority(self, priority):
		self.priority = priority

	def getPriority(self):
		return self.priority

	def setStatus(self, event, *payload):
		self.log.append([datetime.now(), event, payload])

	def getLog(self):
		return self.log

	def getLogString(self):
		log_string = ""
		for entry in self.log:
			timestamp = entry[0].isoformat()
			status = TaskStatus(entry[1]).name
			payload = ', '.join([json.dumps(v) for v in entry[2]])
			log_string += f"[{timestamp}] {status}: {payload}\n"
		return log_string

	def setProviderId(self, providerId):
		self.providerId = providerId

	def getProviderId(self):
		return self.providerId

	def complete(self, result):
		self.onCompleted(result)

	def fail(self):
		self.onFailed()

	onCompleted = NOOP
	onFailed = NOOP

	@property
	def id(self):
		return self.task_info['id']

	def getTaskOptions(self):
		return self.task_info.get('task_options')

	def getMaxPrice(self):
		return self.task_info['max_price']

	def getMoneyTimeRatio(self):
		return self.task_info['time_to_money_ratio']
